### PR TITLE
Low-hanging Fruit: UI improvements, off-hand build wand, crafter recipe adds, and expanding raid command capabilities.

### LINKED ISSUE #6231: Stonemason cannot craft polished basalt or polished blackstone
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.4
- MineColonies Version:
- Structurize Version:
- Related Mods and their Versions: none

## Expected behavior
The Stonemason would be able to craft polished basalt or polished blackstone, since they're made from 4 basalt and 4 blackstone respectively, and both basalt and blackstone can be considered stone.

## Actual behavior
The recipes cannot be taught to the stonemason.

## Steps to reproduce the issue

1. Try to teach the polished blackstone or polished basalt recipes to the stonemason
2. 
3. 

### Logs

- latest.log: 
- crashlog: 

### Notes
Reported on Discord

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #6232: Build Tool can not be used from left hand in survival while Minecolonies is installed.
### Prerequisites 

- [x] I am running the latest **alpha** version of Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.4
- Structurize Version: 1.13.102-ALPHA
- MineColonies Version (if related bug): 1.13.510-APLHA
- Related Mods and their Versions: Minecolonies Official Modpack 2.1

## Expected behavior

Expect to be able to use the build tool in left hand as well as right hand.

## Actual behavior

When I right click on a solid block with Build tool in left hand, the GUI pops up and disappears and I get a chat message (see image bellow.

Build Tool Bug

## Steps to reproduce the issue

1. Use Official MineColonies Modpack
2. Put Build tool in off-hand (left hand)
3. Right click on a solid block.

### Logs

- latest.log: 
- crashlog: 

### Notes

I obviously have Minecolonies installed.
---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### PR DESCRIPTION
**Pull request by gattsuru**
Closes #6231
Closes #6232
Closes https://github.com/ldtteam/minecolonies-features/issues/176
Closes https://github.com/ldtteam/minecolonies-features/issues/200
Closes https://github.com/ldtteam/minecolonies-features/issues/212
Closes https://github.com/ldtteam/minecolonies-features/issues/215
Closes https://github.com/ldtteam/minecolonies-features/issues/222

- Adds a progress message at 25 citizens, pointing players to the University research.
- Adds disabled/enabled button functionality and tooltips for Warehouse Upgrade and Sort.
- Adds a time check to WAREHOUSE_FULL messages, to reduce message spamming, and gives different messages depending on whether the warehouse can still be upgraded (in which case it can't use emerald blocks) or not (in which case it can't be upgraded.
- Adds glowstone dust as an input to the mechanic, and sets glowstone (block) as non-reducible to avoid potential surprises to datapack makers. Added a few other exclusions for storage-like blocks that Forge does not inherently tag as storage, for similar reasons.
- Adds Blackstone and Polished Blackstone (and polished blackstone slab/bricks) and Polished Basalt as usable ingredients to the Stonemason.
- Allows use of build tool in off-hand.
- Removes duplicate sendPlayersMessage for WorkOrderBuildRemovals.
- Adds a message on completion of a deconstruction work order to warn players to use the Pick Up botton.
- Add a third count to the Town Hall's citizen summary page, indicating the maximum limit of colonists from research.
- Adds an optional argument to raid-now, raid-now-all, raid-tonight, and raid-tonight-all to force the raid to be a specific type.
- Adds a raid event for pirates without a ship, only summonable by manual raid commands for now.

Review please

### CODE REVIEW

File: src/main/resources/assets/minecolonies/lang/en_us.json
Comment (ravenbuilder934) [inline src/main/resources/assets/minecolonies/lang/en_us.json]:
```suggestion
  "com.minecolonies.coremod.gui.warehouse.upgrade.disabled.1": "Rack upgrade requires",
```

File: src/main/resources/assets/minecolonies/lang/en_us.json
Comment (ravenbuilder934) [inline src/main/resources/assets/minecolonies/lang/en_us.json]:
```suggestion
  "com.minecolonies.coremod.progress.twenty_five_citizens_spawned": "§6 Your colony's population has grown to the limits of current management techniques. To accept more citizens, complete the research at the University.",
```

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java]: sortButton.disable and add disable texture to xml

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:126]: probably might be worth using \n instead, I also think that tooltip renderer has auto wrapping but not sure about that

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java]: variable name

**Review by Nightenom — CHANGES_REQUESTED** [review]
Few formatting mistakes, try to look at gui system - there are visible and enable mechanisms

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java
Comment (gattsuru) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:126]: Is there a particular trick to getting translation text components to like \n?  I couldn't figure out how to escape them being parsed.

I did see the underlying tooltip rendering's autowrapping, but I couldn't find how to toggle it on; Pane and descendents pretty consistently use renderToolTip over renderWrappedToolTip.

**Review by gattsuru — COMMENTED** [review]
Not sure if there's a trick to getting \n to play nice with tooltips.  The rest of the requested changes are set.

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:126]: No, this is fine.

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java]: While you're on this, there is a small bug here that was reported, citizensSize is also the count limit, instead of the total number of existing citizens

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java]: just use the world ticks here for timing I'd say.

**Comment by ravenbuilder934** [comment]
Did you check with the jungle creator to make sure the concrete's not intentional? Can you send screenshots of the before and after?

File: src/main/resources/assets/minecolonies/lang/en_us.json
Comment (ravenbuilder934) [inline src/main/resources/assets/minecolonies/lang/en_us.json]:
```suggestion
  "com.minecolonies.coremod.gui.townHall.population.totalcitizens.researchlimited": "Research at University",
```

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java
Comment (Mekle001) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:134]: So, emerald block makes sense at level 5, but a warehouse full at level 4 this message is confusing.

File: src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json
Comment (Mekle001) [inline src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json:11]: If we're allowing compression to blocks here, do we also want to allow decompression, and add the glowstone block?

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java
Comment (gattsuru) [reply src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:134]: It should only give the "The Warehouse is full, please use an emerald block to make more space in the racks!" message at level five.  At level four, it will give "The Warehouse is full, please upgrade it!".  Did I goof that up, or does the level four or lower message need different wording?

I guess it might make sense to add a third message for when the warehouse is T5 and has all of its emerald upgrades, if that's the concern.

File: src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json
Comment (gattsuru) [reply src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json:11]: It's possible, but that almost feels more of a Crusher job, especially since the Crusher's better equipped to give better output rates as the building is upgraded.

Up to you all, of course.

Comment (Mekle001) [reply src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json:11]: Yeah, I agree that the crusher might be the best place. If I remember correctly, vanilla doesn't have a block -> dust recipe, but  many modpacks do. If we enable the recipe at the crusher, it will be the first 'taught' recipe that the crusher has, unless we want to enable this as a custom recipe that allows it in vanilla. That last I'd like to hold off on, as that starts to go down the path of non-deterministic inputs to output ratios, which probably won't be supported for a while.

File: src/api/java/com/minecolonies/api/util/constant/Constants.java
Comment (Nightenom) [inline src/api/java/com/minecolonies/api/util/constant/Constants.java]: Small nitpick if you are going to push something else before merge:
This value should be something like `TICKS_SECOND * SECONDS_A_MINUTE * 5`
You can fix the above too

File: src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java]: This method and the command build seems to be dupe code to me, might be worth to merge "raid now" and "raid tonight" into one command with a boolean flag.
At least try to make common logic abstract.

File: src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java]: This should be probably handled from some kind of registry or at least a list of raid types

**Review by Nightenom — DISMISSED** [review]
Gui-wise it's ok for me
Unless someone else agrees on the abstraction of raid commands, it's ok for me

File: src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java
Comment (gattsuru) [reply src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java]: If a registry, any preferences on where to put it?  It's not exactly com.minecolonies.api.entities or com.minecolonies.api.colony.colonyEvents, but feels a little ostentatious making a separate package like com.minecolonies.api.colony.raidTypes to carry five strings.

**Comment by gattsuru** [comment]
Fixed the time constants, set up a RaidTypes registry.  The registry is very sparse, but it's a lot of philosophical and architectural questions to expand to things like mob entities or sounds, and I wanted to keep this light for now.

Unified the -now and -tonight variants of each of the raid-All and raid colony commands, cleaned up the code inside them.  They use the registries, as do the corresponding checks in the raid manager.  Sorry for taking the dumbest approach to populating a string[].  I took a look at merging raid and raid-all, but I don't know that it'd be a good plan with how Brigadier works, and there's less overlap in the function than expected.

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/6288/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

**Comment by someaddons** [comment]
Hm I'm not too happy with the additional registry for raids, which is completly desynced of the actual raidevents registry, which are found in the colonyeven registry. Now we got "RaidTypes" which actually are just a string as seperate registry and are only linked through a string comparison in the raidmanager

**Review by someaddons — APPROVED** [review]

**Comment by ravenbuilder934** [comment]
image

Also, it says `norsemen_raid`, `norsemen_ship_raid`, and `pirate_raid` (with a few others). Does it mean `pirate_raid`, `pirate_ship_raid`, and `norsemen_raid`?

When I triggered a `norsemen_ship_raid`, the norsemen spawned in the water without a ship.

**Comment by gattsuru** [comment]
Ugh, sorry. "babarian"'s part of the underlying ColonyEventTypes registry, will fix that.

((If someone gives a typo, rather than not adding anything in the optional field (which results in normal biome-dependent spawns), they'll end up with barbarians, so it's 'just' a problem with the suggestion builder.  But still something that should be fixed.))

Was the water spawn large enough for their ship at the given colony size? I don't overrule the normal checks for ship positions, so a "norsemen_ship_raid" (and "pirate_raid") won't spawn a ship if there's not an acceptable body of water.  With how raid manager works, I wasn't sure we'd want to do a search for valid spawn locations in the event of no matching result, and it seemed better to drop the land variant of the event instead in that situation.

**Comment by ravenbuilder934** [comment]
I think it was big enough, my test colony was on an island in the middle of an ocean

**Pull request merged**
